app/src/main/java/net/aiscope/gdd_app/model/SampleModels.kt

```diff
@@ -56,6 +53,54 @@ data class MicroscopeQuality(
     val magnification: Int
 )
 
+data class Images(
+    val inProgressCapture: InProgressCapture?,
+    val completedCaptures: List<CompletedCapture>
+) {
+
+    constructor() : this(null, emptyList())
+
+    fun newCapture(path: File): Images = Images(InProgressCapture(path), completedCaptures)
+
+    fun addMask(maskPath: File, isEmpty: Boolean): Images {
+        val existingMaskIndex = completedCaptures.indexOfFirst { it.mask == maskPath }
+
+        return when {
+            existingMaskIndex >= 0 && inProgressCapture != null ->
+                // we somehow broke the business logic
+                throw IllegalStateException(
+                    "Trying to add mask $maskPath at index $existingMaskIndex " +
+                            "when inProgressCapture is not null ($inProgressCapture)"
+                )
+            inProgressCapture != null ->
+                Images(
+                    null,
+                    completedCaptures + CompletedCapture(
+                        inProgressCapture.image, maskPath, isEmpty
+                    )
+                )
+            else ->
+                Images(
```

Comment:
Sorry, what is this use case exactly?

Reply:
This branch is executed when editing the mask of a previous "capture". The code is not very clear about that, so I added some tests which I missed adding in the first place, to serve as documentation, and also did some refactoring and added some comments. Let me know what you think now, probably there is room for more improvement. See 91481252.